Lesson19/didem/script.js

```diff
@@ -0,0 +1,176 @@
+-----------------------------------------------------------
+1. Define a `ShoppingCart` class.
+2. Add a constructor that initializes an empty private 
+   array `#items` to store the cart items.
+3. Add a `viewCart` method to display all items in the cart.
+*/
+
+/*
+-----------------------------------------------------------
+  STEP 2: Add Items to the Cart
+-----------------------------------------------------------
+1. Create an `addItem` method in the `ShoppingCart` class.
+2. The method should:
+   - Accept `name`, `price`, and `quantity` as parameters.
+   - Check if the item already exists in the cart.
+     - If it exists, increase the quantity.
+     - Otherwise, add the new item to the `#items` array.
+*/
+
+/*
+-----------------------------------------------------------
+  STEP 3: Remove Items from the Cart
+-----------------------------------------------------------
+1. Add a `removeItem` method to the `ShoppingCart` class.
+2. The method should:
+   - Accept the `name` of the item to remove.
+   - Remove the item from the `#items` array if it exists.
+*/
+
+class ShoppingCart {
+  #items;
+
+  constructor() {
+    this.#items = [];
+  }
+
+  viewCart() {
+    console.log('Viewing cart....');
+    for (const item of this.#items) {
+      console.log(item);
+    }
+  }
+
+  addItem(name, price, quantity) {
+    console.log('Adding new item to the cart...');
+    for (const item of this.#items) {
+      if (name === item.name) {
+        console.log('increase the quantity.');
+        item.quantity++;
+        return;
+      }
+    }
+
+    const id = new Date().getMilliseconds();
+    this.#items.push({ id, name, price, quantity });
+  }
+
+  // In real life, we usually use id to find elements or do any manipulations
+  removeItem(name) {
+    for (let i = 0; i < this.#items.length; i++) {
+      const currentElement = this.#items[i];
+      if (name.toLowerCase() === currentElement.name.toLowerCase()) {
+        this.#items.splice(i, 1);
+        return;
+      }
+    }
+  }
+getTotal() {
+    let totalAmount = 0;
+    // Sepetteki her öğe için döngü
+    for (const item of this.#items) {
+      // Toplam fiyata (fiyat * miktar) ekle
+      totalAmount += item.price.amount * item.quantity;
+    }
+    // Para birimini (currency) ilk öğeden al, yoksa 'USD' varsay
+    const currency = this.#items.length > 0 ? this.#items[0].price.currency : 'TRY';
+    
+    // Toplamı ve para birimini içeren bir obje döndür
+    return { amount: totalAmount, currency: currency };
+  }
+
+applyDiscount(code) {
+    const validDiscounts = {
+      'SAVE10': 0.10, 
+      'SAVE20': 0.20, 
+      'KURS30': 0.30  
+    };
+
+    if (!validDiscounts[code]) {
+      console.log(`Error: Invalid discount code: ${code}`);
+      return this.getTotal(); 
+      }
+    const discountRate = validDiscounts[code];
+    const originalTotal = this.getTotal();
+    
+    const discountedAmount = originalTotal.amount * (1 - discountRate);
```

Comment:
This code works only if `price` is an object with `amount` and `currency` fields, but we didn’t define `price` that way in Step 2. Since `getTotal()` will return a number, this should be:
```suggestion
    const discountedAmount = originalTotal * (1 - discountRate);
```